src/resources/lib/modules/system.py

```diff
@@ -341,13 +340,59 @@ def exit(self):
         self.oe.dbg_log('system::exit', 'exit_function', 0)
         pass
 
+    # Identify connected GPU card (card0, card1 etc.)
+    def get_gpu_card(self):
+        card = None
+        for root, dirs, files in os.walk("/sys/class/drm", followlinks=False):
+            for dir in dirs:
+                try:
+                    with open(os.path.join(root, dir, 'status'), 'r') as infile:
+                        for line in [x for x in infile if x.replace('\n', '') == 'connected']:
+                            card = dir.split("-")[0]
+                except:
+                    pass
+                if card: break
+            break
+
+        if card:
+          return card
+        else:
+          return 'card0'
+
+    def get_gpu_type(self):
+        if not self.oe.ARCHITECTURE.endswith('.x86_64'):
+            self.oe.dbg_log('system::get_gpu_type', 'Architecture is %s, assume default SOC GPU' % self.oe.ARCHITECTURE, 0)
+            return 0
+
+        gpu_drivers = {'i915': 1, 'i965': 2,
+                       'nvidia': 3, 'nvidia-legacy': 4,
+                       'r300': 5, 'r600': 6, 'radeonsi': 7,
+                       'vmwgfx': 8, 'virtio_gpu': 9}
+
+        gpu_card = self.get_gpu_card()
+        self.oe.dbg_log('system::get_gpu_type', 'Using card: %s' % gpu_card, 0)
+
+        if gpu_card:
+            gpu_driver = self.oe.execute('/usr/bin/udevadm info --name=/dev/dri/%s --attribute-walk 2>/dev/null | /usr/bin/grep -m1 DRIVERS= | /usr/bin/sed \'s/["=]/ /g\' | /usr/bin/awk \'{print $2}\'' % gpu_card, get_result=1)
```

Comment:
this is pretty ugly, can we not just get the info from udevadm and parse in python?

Reply:
Yes it's ugly, but the output of `udevadm` isn't particularly amenable to Python processing - it's not JSON, unfortunately:

```
LibreELEC:~ # udevadm info --name=/dev/dri/card0 --attribute-walk

Udevadm info starts with the device specified by the devpath and then
walks up the chain of parent devices. It prints for every device
found, all possible attributes in the udev rules key format.
A rule to match, can be composed by the attributes of the device
and the attributes from one single parent device.

  looking at device '/devices/pci0000:00/0000:00:1c.0/0000:01:00.0/drm/card0':
    KERNEL=="card0"
    SUBSYSTEM=="drm"
    DRIVER==""

  looking at parent device '/devices/pci0000:00/0000:00:1c.0/0000:01:00.0':
    KERNELS=="0000:01:00.0"
    SUBSYSTEMS=="pci"
    DRIVERS=="nvidia"
    ATTRS{boot_vga}=="1"
    ATTRS{broken_parity_status}=="0"
    ATTRS{class}=="0x030000"
    ATTRS{consistent_dma_mask_bits}=="32"
    ATTRS{current_link_speed}=="2.5 GT/s"
    ATTRS{current_link_width}=="1"
    ATTRS{d3cold_allowed}=="1"
    ATTRS{device}=="0x0a64"
    ATTRS{dma_mask_bits}=="40"
    ATTRS{driver_override}=="(null)"
    ATTRS{enable}=="2"
    ATTRS{irq}=="30"
    ATTRS{local_cpulist}=="0-3"
    ATTRS{local_cpus}=="f"
    ATTRS{max_link_speed}=="5 GT/s"
    ATTRS{max_link_width}=="16"
    ATTRS{msi_bus}=="1"
    ATTRS{numa_node}=="-1"
    ATTRS{revision}=="0xa2"
    ATTRS{subsystem_device}=="0x047a"
    ATTRS{subsystem_vendor}=="0x1025"
    ATTRS{vendor}=="0x10de"

  looking at parent device '/devices/pci0000:00/0000:00:1c.0':
    KERNELS=="0000:00:1c.0"
    SUBSYSTEMS=="pci"
    DRIVERS=="pcieport"
    ATTRS{broken_parity_status}=="0"
    ATTRS{class}=="0x060400"
    ATTRS{consistent_dma_mask_bits}=="32"
    ATTRS{current_link_speed}=="2.5 GT/s"
    ATTRS{current_link_width}=="1"
    ATTRS{d3cold_allowed}=="1"
    ATTRS{device}=="0x27d0"
    ATTRS{dma_mask_bits}=="32"
    ATTRS{driver_override}=="(null)"
    ATTRS{enable}=="1"
    ATTRS{irq}=="24"
    ATTRS{local_cpulist}=="0-3"
    ATTRS{local_cpus}=="f"
    ATTRS{max_link_speed}=="2.5 GT/s"
    ATTRS{max_link_width}=="1"
    ATTRS{msi_bus}=="1"
    ATTRS{numa_node}=="-1"
    ATTRS{revision}=="0x02"
    ATTRS{secondary_bus_number}=="1"
    ATTRS{subordinate_bus_number}=="1"
    ATTRS{subsystem_device}=="0x047a"
    ATTRS{subsystem_vendor}=="0x1025"
    ATTRS{vendor}=="0x8086"

  looking at parent device '/devices/pci0000:00':
    KERNELS=="pci0000:00"
    SUBSYSTEMS==""
    DRIVERS==""

```